### ISSUE TITLE
Search tree module

### ISSUE THREAD

**Issue by EmilyOng** [EmilyOng/AlgebraicEffect#3]
Sample program: [`foldr_length`](https://github.com/EmilyOng/AlgebraicEffect/blob/StagedSL/src/examples/fold.ml#L2-L17)

**Note 1**

Proof is currently at: `[any; unfold rhs] ex v92 v93; req emp; ens tail(xs)=v93/\is_cons(xs)=true; length(v93, v92); ex r; req emp; Norm(r=1+v92/\res=r+init)`

![Screenshot 2023-12-16 at 11 37 06 AM](https://github.com/EmilyOng/AlgebraicEffect/assets/21200681/2db35246-2531-4e9d-aa90-02c9105c816b)

Now, the proof succeeds but `all` does not seem to be respected (because it has a failed `any` child, hidden due to compact search tree representation). Shouldn't it be `✘`?

![Screenshot 2023-12-16 at 11 37 19 AM](https://github.com/EmilyOng/AlgebraicEffect/assets/21200681/700bd1ad-a2f5-4dca-b08c-e3497627ecca)

**Note 2**

The `✔ [all; unfold lhs] ex v53; req emp; Norm(is_nil(xs)=true/\res=init)` flow is marked to succeed though it has an unverified `any`. Below is a non-compact tree representation.

![Screenshot 2023-12-16 at 11 53 22 AM](https://github.com/EmilyOng/AlgebraicEffect/assets/21200681/70ce423e-d366-4fc0-89df-18fbff3cea04)

(tracing) State update over here https://github.com/EmilyOng/AlgebraicEffect/blob/StagedSL/parsing/search.ml#L80

**Comment by EmilyOng** [comment]
Another question for [`take_skip`](https://github.com/EmilyOng/AlgebraicEffect/blob/StagedSL/src/examples/later/take_skip.ml#L31-L49)

`[any; disj rhs] Norm(res=[])` is marked as failed prematurely, should the subsequent flows still be tried?

![Screenshot 2023-12-16 at 11 41 43 AM](https://github.com/EmilyOng/AlgebraicEffect/assets/21200681/775b3701-7ff5-4b68-9a27-5a40430e0ad9)

**Comment by dariusf** [comment]
"any" means that the _parent_ goal succeeds if any child succeeds. In other words, _any_ of the nodes marked "any" have to succeed.

I went through all 3 examples and think they are correct under this interpretation. For example, in the last one, the "any" has failed because _all_ its children have to succeed, which is no longer possible.

Anyway, I agree that it's confusing. I've tweaked the drawing, hopefully it's more intuitive (and less noisy) now. If not we can always change it back/further.
